### PR TITLE
add fallback for max pod calculation

### PR DESCRIPTION
**Pull request by Issacwww**
**Issue #, if available:**

**Description of changes:**
calculate max pod when instance type is not present using EC2 client

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Testing Done**
`make test && make test-e2e`

*[See this guide for recommended testing for PRs.](../doc/CONTRIBUTING.md#testing-changes) Some tests may not apply. Completing tests and providing additional validation steps are not required, but it is recommended and may reduce review time and time to merge.*

### CODE REVIEW

File: nodeadm/go.mod [hunk 1]
```diff
@@ -22,7 +22,9 @@ require (
 
 require (
 	github.com/Microsoft/go-winio v0.6.1 // indirect
+	github.com/aws/aws-sdk-go-v2/service/ec2 v1.146.0 // indirect
```
Comment (ndbaker1) [inline nodeadm/go.mod:25]: does aws sdk add 400k lines? 😱

File: nodeadm/internal/kubelet/config.go [hunk 2]
```diff
@@ -263,16 +260,17 @@ func (ksc *kubeletConfig) withCloudProvider(cfg *api.NodeConfig, flags map[strin
 func (ksc *kubeletConfig) withDefaultReservedResources(cfg *api.NodeConfig) {
 	ksc.SystemReservedCgroup = ptr.String("/system")
 	ksc.KubeReservedCgroup = ptr.String("/runtime")
-	memoryReservation := getMemoryMebibytesToReserve(defaultMaxPods)
+	zap.L().Info(fmt.Sprintf("calculate the max pod for instance type: %s", cfg.Status.Instance.Type))
```
Comment (cartermckinnon) [inline nodeadm/internal/kubelet/config.go:263]: You should use a zap field for the instance type is read of using Sprintf here

File: nodeadm/go.mod (hunk 1 above)
Comment (Issacwww) [reply nodeadm/go.mod]: and also `github.com/stretchr/objx`?

File: nodeadm/internal/kubelet/eni_max_pods.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+// https://github.com/awslabs/amazon-eks-ami/blob/master/files/bootstrap.sh#L514
+// which essentially is
+//
+//	# of ENI * (# of IPv4 per ENI - 1) + 2
+func CalcMaxPods(awsRegion string, instanceType string) int32 {
+	zap.L().Info("calculate the max pod for instance type", zap.String("instanceType", instanceType))
+	cfg, err := config.LoadDefaultConfig(context.Background(), config.WithRegion(awsRegion))
+	if err != nil {
+		zap.L().Warn("error loading AWS SDK config when calculating the max pod, setting it to default value",
+			zap.Int("defaultMaxPods", defaultMaxPods),
```
Comment (ndbaker1) [inline nodeadm/internal/kubelet/eni_max_pods.go:59]: do we need to log the default? just think its adding space when users can find this from resulting config

Comment (Issacwww) [reply nodeadm/internal/kubelet/eni_max_pods.go]: Fair enough, removed this field

**Review by ndbaker1 — APPROVED** [review]

File: nodeadm/internal/util/ec2_test.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+				InstanceTypes: []types.InstanceTypeInfo{},
+			},
+			mockError:     fmt.Errorf("invalid instance type"),
+			expectedError: fmt.Errorf("error describing instance type mock-type.large: %w", errors.New("invalid instance type")),
+		},
+	}
+
+	for _, test := range tests {
+		mockEC2 := &MockEC2Client{}
+		mockEC2.On("DescribeInstanceTypes", mock.Anything, mock.AnythingOfType("*ec2.DescribeInstanceTypesInput")).Return(&test.mockResponse, test.mockError)
```
[the remark this answers is not in the document]
Comment (Issacwww) [reply nodeadm/internal/util/ec2_test.go:74]: not get this question, what library are you expecting to see?

[the remark this answers is not in the document]
Comment (ndbaker1) [reply nodeadm/internal/util/ec2_test.go:74]: like just writing `MockEC2Client`'s constructor to take values to return when calling `DescribeInstanceTypes`. the mocks make sense for an interface deeply nested within other code, but this is just calling it directly 🤷

[the remark this answers is not in the document]
Comment (Issacwww) [reply nodeadm/internal/util/ec2_test.go:74]: Ack, will do a clean up in next PR

**Review by cartermckinnon — APPROVED** [review]

**Pull request merged**